corehq/apps/users/views/utils.py

```diff
@@ -115,58 +116,87 @@ def log_commcare_user_locations_changes(request, user, old_location_id, old_assi
-            .size(0))
-    ).run()
-    locations = SQLLocation.objects.filter(location_id__in=location_ids)
     counts = query.aggregations.by_location.counts_by_bucket()
-    return locations, counts
+    for location_id in location_ids:
+        if location_id not in counts:
+            counts[location_id] = 0
+
+    return counts
 
 
-def get_locations_with_single_user(domain, location_ids, user_id):
+def get_locations_with_orphaned_cases(domain, location_ids, user_id):
     """
     Takes a list of location IDs and returns all the given location IDs
-    where the user ID is the only one that has cases there.
+    where the user ID is the only one that has cases there. Also searched through descendant locations
+    if a location has `view_descendants` enabled.
     :param domain: The domain to search in.
     :param location_ids: A list of location IDs to get case counts on.
-    :param user_id: The user ID to check given location_ids with as the only assigned user with cases.
+    :param user_id: The user ID to check given `location_ids` with as the only assigned user with cases.
     :returns A dict with location names as the key, and the number of cases assigned to them as the value.
     """
-    # Get locations where only given user_id has cases in list of locations
-    location_ids_with_single_user = _get_location_ids_with_single_user(domain, location_ids, user_id)
-    locations_with_single_user = dict()
-    if location_ids_with_single_user:
-        locations, counts = _get_location_case_counts_with_single_user(domain, location_ids_with_single_user)
-        for location in locations:
-            if location.location_id in counts:
-                locations_with_single_user[location.name] = counts[location.location_id]
-            else:
-                locations_with_single_user[location.name] = 0
-
-    return locations_with_single_user
+    loc_ids_with_other_users = _get_location_ids_with_other_users(domain, location_ids, user_id)
+
+    # Get the assigned location ids that are in all_loc_ids but not in loc_ids_with_other_users
+    loc_ids_with_orphaned_cases = list(set(location_ids) - loc_ids_with_other_users)
+    if loc_ids_with_orphaned_cases:
+        assigned_locs = SQLLocation.objects.filter(location_id__in=loc_ids_with_orphaned_cases)
+        # We need to find all descendant locations for orphaned assigned locations that don't have a user
+        # assigned to them
+        desc_locs = SQLLocation.objects.get_queryset_descendants(
+            assigned_locs.filter(location_type__view_descendants=True)
+        )
+        desc_loc_ids = [loc.location_id for loc in desc_locs]
+        desc_loc_ids_with_other_users = _get_location_ids_with_other_users(domain, desc_loc_ids, user_id)
+        desc_loc_ids_with_orphaned_cases = list(set(desc_loc_ids) - desc_loc_ids_with_other_users)
+
+        all_orphaned_loc_ids = set(loc_ids_with_orphaned_cases + desc_loc_ids_with_orphaned_cases)
```

Comment:
I think you could simplify this code a bit by getting the list of all assigned locations and relevant descendants up front, then checking for users on that complete set.  The way it's done here involves two calls to `_get_location_ids_with_other_users` and associated code.  I guess this way pre-filters the set of locations to descend into, which is valid, is that what you were thinking about?

Reply:
I tried implementing it so that everything is fetched in the beginning, but the difficulty with that is that we need to only check descendants of assigned locations that are at risk of being orphaned. Not doing this, one could end up where multiple users share a location with `view_descendants` enabled, but then all those users will see the descendants as being at risk of becoming orphaned.

Reply:
I think this is also true if any ancestor locations view descendants and have users.  Like, if your location hierarchy is:

* Massachusetts
  * Suffolk County
    * Boston
    * Chelsea
  * Middlesex County
    
Imagine there are cases in Boston, and a Suffolk County user is reassigned to Middlesex.  If there are no other users at Suffolk, and none in Boston, then this code will correctly identify those cases as being abondoned.  However, if there's a user at the Massachusetts level, then they won't actually be at risk, but this isn't being taken into account.

Reply:
That is true, the code isn't considering ancestor locations. One would need to go through all the ancestors that have `view_descendants` enabled and then check if there are any users assigned to these locations. What stopped me from implementing this was the fact that it might not always be relevant, as users assigned to ancestor locations might not be directly dealing with cases from descendant locations. Nevertheless, would this be something that warrants implementing?

Reply:
> would this be something that warrants implementing?

I don't know - at a high level I do fear that this whole location alert thing is overly specific to a particular workflow, though it does sound like a helpful note when applicable.

I think if it factors in cases potentially abandoned at descendant locations, then it makes sense to consider users who are ancestors of those locations too (given that the user being unassigned is potentially also at an ancestor location).  I guess the simplest approach would be to ignore ancestors and descendants, alert only about cases at locations to which the user is directly assigned, and accept that this alert might not be perfectly accurate if the project has a workflow where cases can be operated on by users at ancestor locations.  You'd probably want to work on very clear language in the alert for that, of course.

After working through the difficulty of getting anything comprehensive, I'm inclined to got to that approach, especially since there are other ways cases might still be visible to other users, such as extension cases.

Reply:
That is a good point, and I feel like this can get very complex when trying to accommodate all possible scenarios. I do feel that, since this is only an alert to notify the user, we can lean towards the more simple approach and then expand upon it from there if it becomes a need in the future. 

I have already implemented checking descendant locations, so I will leave that code in there as it gives a slightly more comprehensive notification, but I will then rewrite the message to indicate that it is only for descendant locations and that ancestors have not been checked.